### ISSUE TITLE
Prometheus' /query endpoint returning data After requested time

### ISSUE THREAD

**Issue by jacksontj** [prometheus/prometheus#5972]
## Bug Report
**What did you do?**
I sent a query at timestamp `1567092749.3346083`

**What did you expect to see?**
I expected to see a result with a timestamp of `1567092749.3346083`

**What did you see instead? Under which circumstances?**
I saw a result with a timestamp of `1567092749.335` -- which is technically *after* my requested time. I am able to reproduce this on master as well-- the issue comes from [here](https://github.com/prometheus/prometheus/blob/master/web/api/v1/api.go#L1211) where the parseTime actually rounds the nanoseconds. This is specifically an issue in [promxy](https://github.com/jacksontj/promxy/issues/212) as this results in data *after* the requested time which the promql engine will ignore (generally it handles data that is at or before the time, not after).

https://play.golang.org/p/VYCV1JFPIKN showcases the issue. Go playground doesn't do time properly but if you run it locally you'll get output like:

```
old
1567093768.0287952
2019-08-29 08:49:28.029 -0700 PDT
1567093768.029
new
1567093768.029
2019-08-29 08:49:28.029 -0700 PDT
1567093768.029
```

So I see 2 paths forward (1) remove the rounding from parseTime or (2) change the rounding to truncate. I'm happy to provide a patch for either, I just need to know which approach we prefer.

**Comment by brian-brazil** [comment]
Prometheus only handles milliseconds, so providing more precision is undefined really.

**Comment by jacksontj** [comment]
true, the pain point is that someone can send a time (as described) and get an invalid response (data in the future). So for using the prom UI this is a non-issue, but of anyone is using the API and sends higher-precision it can return an invalid response. IMO prom should either (1) truncate or (2) error the way it is now is just confusing :/

**Comment by brian-brazil** [comment]
I don't think I'd class the response as invalid, and the question is more what the expected behaviour is with Javascript which only supports floats.

**Comment by jacksontj** [comment]
I'd classify it as "invalid" only because the data returned is for a time strictly after the time requested (it should match -- not be before or after)

e.g. 1567093768.0287952 shouldn't ever return 1567093768.029

Which is why I think truncating or erroring would make more sense.

**Comment by brian-brazil** [comment]
> it should match -- not be before or after

That's not possible, if you provide more precision than milliseconds then it's always going to violate one of those. Truncating doesn't help, and erroring would run afoul of Javascript and be a bit user hostile.

**Comment by jacksontj** [comment]
True, but rounding is IMO worse than truncation because its either before **or** after instead of being consistent. Especially if you wire the API up to the promql engine it is *very* strict about not taking data after the requested time.

**Comment by brian-brazil** [comment]
It's consistent, and makes no practical difference for graphing.
The query API is not a storage API so that's not going to work well for other reasons, and in any case the engine can't request anything other than milliseconds.

**Comment by jacksontj** [comment]
The engine can request more than Ms, internally lookups are just time.Time
which allows more precision (I had this linked issue using the promql
engine).

**Comment by brian-brazil** [comment]
It can't, it's all millis: https://github.com/prometheus/prometheus/blob/cb5213e91438d0a0341631f8049c6c637d394a9b/storage/interface.go#L49

**Comment by jacksontj** [comment]
The engine is time.Time (
https://github.com/prometheus/prometheus/blob/master/promql/engine.go#L311)
which means somewhere it's getting translated, so maybe the same problem at
that level. I'm addition the go client is time.Time (
https://github.com/prometheus/client_golang/blob/master/api/prometheus/v1/api.go#L236)
I can submit a pr to truncate in there

**Comment by brian-brazil** [comment]
What I am saying is that PromQL can only ask for millis from storage, so I'm not seeing any actual problem here.

**Comment by jacksontj** [comment]
The issue is the conversion. If a time is sent that's too precise the web
rounds and sends millis to the storage, which as previously mentioned can
lead to results which are strictly after the requested time -- hence a
problem

**Comment by krasi-georgiev** [comment]
I haven't been involved with the API usages  at all so my answer is just a guess, but it seems that error-ing is the right approach here. I mean if the request is not with the expected values the API should return an error (or a warning if that can be part of the API response.) 

> erroring would run afoul of Javascript and be a bit user hostile.

@brian-brazil can you give an example why you think this?

**Comment by brian-brazil** [comment]
Javascript has no support for integers, so users can't sanely control if they end up with a value that's not an exact millisecond value.

**Comment by krasi-georgiev** [comment]
How does this correlate with error-ing or warning when making a request with a higher than the supported precision?

**Comment by brian-brazil** [comment]
Floating point inaccuracy would cause unexpected errors.

**Comment by krasi-georgiev** [comment]
Than instead of an error can return the results along with just a warning.

**Comment by brian-brazil** [comment]
I'm not sure this is an appropriate use of a warning, especially as some users will treat warnings as fatal errors.

**Comment by krasi-georgiev** [comment]
But this is their choice right? The api returns the results and they can decide what to do with it.

**Comment by krasi-georgiev** [comment]
@jacksontj would that work for your use case?

**Comment by jacksontj** [comment]
> Javascript has no support for integers, so users can't sanely control if they end up with a value that's not an exact millisecond value.

If this is the case maybe we should make this behavior client dependant? e.g. have the client_golang client error, but the javascript side won't? That seems somehow worse though

> @jacksontj would that work for your use case?

Unfortunately not, the `warnings` today are only storage level warnings and specifically are treated as non-fatal (IMO they seem very much like an afterthought and still have a lot of kinks to work out). In addition they are just strings so I wouldn't be able to reliably determine that one of the warnings was "time is too granular". Even if we solved those it wouldn't solve the real issue where I'd get data *after* my request -- I still have to truncate the time somewhere else. In my use-case I just wrapped the prom API to always truncate the time to milliseconds so I won't get this "data after request" issue.
